### PR TITLE
near-store: extend state dump file to include FlatState

### PR DESCRIPTION
**Pull request by mina86**
Extend state dump to include FlatState column.  State dump format is
changed to support multiple columns by including column index along
with every key-value pair.  This introduces an overhead of 1 byte per
entry which should be negligible.

While we are changing the format, also add an end of file mark so that
we can properly detect truncated files.  Previously, if we were very
unlucky and the file was truncated just at a record border we wouldn’t
notice it when reading the file.  With explicit end of file marker we
can detect it now.

Co-authored-by: Anton Puhach <anton@near.org>
Issue: https://github.com/near/nearcore/issues/8322

### CODE REVIEW

**Comment by Longarithm** [comment]
cc @pugachAG because he also works with these methods rn: https://github.com/near/nearcore/pull/8417
I think that test makes strong sense, and if it doesn't interfere much with work on #8417, I can take a look and approve.

File: core/store/src/lib.rs [hunk 1]
```diff
@@ -325,11 +324,7 @@ impl Store {
         let file = File::create(filename)?;
         let mut file = BufWriter::new(file);
         for item in self.storage.iter_raw_bytes(column) {
-            let (key, value) = item?;
-            file.write_u32::<LittleEndian>(key.len() as u32)?;
-            file.write_all(&key)?;
-            file.write_u32::<LittleEndian>(value.len() as u32)?;
-            file.write_all(&value)?;
+            borsh::to_writer(&mut file, &item?)?
```
Comment (pugachAG) [inline core/store/src/lib.rs:327]: I personally find it a bit confusing to start relying on the fact that `borsh` coincidentally encodes this the same way as we want here. Especially considering that we don't use it when decoding values. I would rather prefer having `write_vec` method (similar to the newly added `read_vec`) to explicitly writes bytes.

**Comment by pugachAG** [comment]
> cc @pugachAG because he also works with these methods rn: #8417 I think that especially a test here makes strong sense, and if it doesn't interfere much with work on #8417, I can take a look and approve.

@Longarithm Thanks for flagging this. It doesn't seem to be in the way of #8417, so we can proceed here and then I will update my implementation. @mina86 you can also include changes from #8417 in this PR if you prefer that.

**Review by pugachAG — APPROVED** [review]
Thanks for including #8417 into this!

**Review by pugachAG — CHANGES_REQUESTED** [review]
Let's not enable nayduck tests as part of this PR. This is not enough to fully enable flat storage as Alex explained in [this comment](https://github.com/near/nearcore/pull/8417#issuecomment-1400534027).

**Comment by mina86** [comment]
Done.

**Review by pugachAG — APPROVED** [review]
LGTM @Longarithm could you please stamp this

**Review by Longarithm — APPROVED** [review]

**Pull request merged**